design-history/exploration-reports/2019.06-unixfsv2-spike-01.md

```diff
@@ -0,0 +1,296 @@
+
+See https://github.com/ipld/specs/issues/130 -- specifically,
+the "referencing" (or "rendezvous") problem -- for more discussion of this.
+
+We're exploring with the "behavioral specification" approach here.
+
+### What are executables, anyway?
+
+In this schema, the executablity bit is in the `Attribs.posix` bitfield...
+as three distinct bools, once per user/group/everyone (the `0111` mask).
+
+Another way to regard executablity which would often (often!) be valid is
+to have a single bool per file.
+
+These are two different opinions application design might take in regarding
+unix*ish* filesystems.  Logically consistent arguments can be made for both.
+Perhaps this is another situation where two different schemas would be useful;
+we can support both, and do so clearly, if we think it's useful to do so.
+
+### Ugh, int sizes
+
+> mtime Int
+
+In designing the IPLD Data Model itself, we still have some deferred decisions
+about how we want to regard integer precision (as well as what we expect,
+in practice, when client implementations in various languages hit their
+host language's natural limitations).
+
+We should probably give particular attention to timestamps in such decisions.
+They're a very common user story.
+
+Additionally, regardless of what precisely we decide to do about integer sizes,
+we should aim to have documentation with recommendations of best practices
+for the timestamp user story.
+
+### Links
+
+The `&` syntax seen in `content &AnyFile` denotes that the `AnyFile` is
+expected on the far side of a Link in that position.
+
+Denotation of the position of a link is important because It Matters To The Hash,
+as well as mattering to the overall understanding of topology of the larger DAG;
+yet it's terse because it doesn't matter to the semantics of the design (much).
+
+The single character sigil would seem to imply that there's no such thing as an `&&Foo` or `&&&Foo`.
+Which indeed I hope there isn't: this author can't think of any situations
+where that expresses a sane thing.  (It's rare enough in programming with pointers;
+it's an entirely separate thing to propose in a merkle-DAG, which has no mutablity
+semantics in the first place, which removes even the 'rare' excuses.)
+We haven't officially explored or documented this anywhere yet.  Probably we should.
+
+### Did you notice we still don't have enough information to determine CIDs?
+
+This schema does not specify which:
+
+- CID version;
+- multicodec;
+- nor multihash
+
+... to use when serializing any block data and constructing links.
+
+It also doesn't specify any restrictions for those values when reading data.
+
+That's... interesting.  What does this imply for work that's still left to do
+in systems which build using these tools?
+
+### Do advanced layouts deserve parameters?
+
+We could imagine having blocks with details such as this in the schema:
+
+```
+advanced Rabin {
+	kind bytes
+	implementation "IPLD/experimental/rabin/v1"
+	bitmatch 14
```

Comment:
I would put the actual values in the root block and having the schema define only the types.

```ipldsch
type RabinParameters struct  {
    bitmatch Int (implicit: 14)
    minchunksize Int (implicit: 256)
    maxchunksize Int (implicit: 262144)
}

advanced Rabin {
    kind bytes
    implementation "IPLD/experimental/rabin/v1"
    parameters RabinParameters
}
```

Reply:
Oh, that's...  wow, that's almost "duh" now that you've said it.  I like that.  We have a system for assigning implicit/default values: it makes perfect sense to use it.

The one tweak I'd make to that is... the `type RabinParameters struct {...}` declaration would make more sense to put *inside the ADL's* schema -- a.k.a. [hereish, in the example of the spec Rod's working on](https://github.com/ipld/specs/pull/131/files#r295781918), I think?  But... hrm, no, that doesn't make it any clearer where another schema that just wants to use an existing thing would configure it from.

Reply:
I like this idea, but also don't intend to try to reach final consensus and resolution on this topic today.  This doc is one example thought.  We should write another to explore this one more fully.

(And as exemplified by how I had to link to a PR comment above, rather than the actual line of interest... I'd still really like to merge this "today"(ish) so it's nicely linkable for those future discussions.)